presto-main/src/main/java/io/prestosql/sql/planner/LocalDynamicFilter.java

```diff
@@ -0,0 +1,211 @@
+                    // See the comment at TupleDomain::columnWiseUnion() for more details.
+                    result = Arrays.stream(partitions).reduce(TupleDomain.none(), TupleDomain::columnWiseUnion);
+                    // Drop collected predicates after collection is over.
+                    partitions = null;
+                }
+            }
+        }
+        if (result != null) {
+            verify(resultFuture.set(convertTupleDomain(result)), "dynamic filter result is provided more than once");
+        }
+    }
+
+    private TupleDomain<Symbol> convertTupleDomain(TupleDomain<String> result)
+    {
+        if (result.isNone()) {
+            return TupleDomain.none();
+        }
+
+        ImmutableMap.Builder<Symbol, Domain> builder = ImmutableMap.builder();
+        for (Map.Entry<String, Domain> entry : result.getDomains().get().entrySet()) {
+            Domain domain = entry.getValue();
+            // Store all matching symbols for each build channel index.
+            for (Symbol probeSymbol : probeSymbols.get(entry.getKey())) {
+                builder.put(probeSymbol, domain);
+            }
+        }
+        return TupleDomain.withColumnDomains(builder.build());
+    }
+
+    public static Optional<LocalDynamicFilter> create(JoinNode planNode, int partitionCount)
+    {
+        Set<String> localDynamicFilters = planNode.getDynamicFilters().keySet();
+        List<FilterNode> filterNodes = PlanNodeSearcher
+                .searchFrom(planNode.getLeft())
+                .where(LocalDynamicFilter::isFilterAboveTableScan)
+                .findAll();
+
+        Multimap<String, Symbol> probeSymbols = MultimapBuilder.treeKeys().arrayListValues().build();
+        for (FilterNode filterNode : filterNodes) {
+            DynamicFilters.ExtractResult extractResult = DynamicFilters.extractDynamicFilters(filterNode.getPredicate());
+            for (DynamicFilters.Descriptor descriptor : extractResult.getDynamicConjuncts()) {
+                if (descriptor.getInput() instanceof SymbolReference) {
+                    // Add descriptors that match the local dynamic filter (from the current join node).
+                    if (localDynamicFilters.contains(descriptor.getId())) {
+                        Symbol probeSymbol = Symbol.from(descriptor.getInput());
+                        log.debug("Adding dynamic filter %s: %s", descriptor, probeSymbol);
+                        probeSymbols.put(descriptor.getId(), probeSymbol);
+                    }
+                }
+            }
+        }
+
+        PlanNode buildNode = planNode.getRight();
+        Map<String, Integer> buildChannels = planNode
+                .getDynamicFilters()
+                .entrySet()
+                .stream()
+                // Add build channels that match local probe symbols.
+                .filter(entry -> probeSymbols.containsKey(entry.getKey()))
```

Comment:
`probeSymbols` are from probe side, right? However, `entry.getKey()` will be symbol from build side, so they will never match. I think you should filter by `localDynamicFilters`

Reply:
I think that I got the comment wrong... instead it should say:
```
// Skip build channels that don't match local probe dynamic filters.        
```
The `probeSymbols` keys are actually probe-side dynamic filter IDs, so the `.filter(...)` above is meant to skip build-side dynamic filters which don't have matching probe-side symbols.
Please let me know if this can be rewritten in a simpler way :)